runners/direct-java/src/main/java/org/apache/beam/runners/direct/StatefulParDoEvaluatorFactory.java

```diff
@@ -244,27 +250,40 @@ public void processElement(WindowedValue<KeyedWorkItem<K, KV<K, InputT>>> gbkRes
         delegateEvaluator.processElement(windowedValue);
       }
 
+      @Nullable Instant lastFired = null;
       for (TimerData timer : gbkResult.getValue().timersIterable()) {
         checkState(
             timer.getNamespace() instanceof WindowNamespace,
             "Expected Timer %s to be in a %s, but got %s",
             timer,
             WindowNamespace.class.getSimpleName(),
             timer.getNamespace().getClass().getName());
-        WindowNamespace<?> windowNamespace = (WindowNamespace) timer.getNamespace();
-        BoundedWindow timerWindow = windowNamespace.getWindow();
-        delegateEvaluator.onTimer(timer, timerWindow);
+        checkState(
+            lastFired == null || !lastFired.isAfter(timer.getTimestamp()),
+            "lastFired was %s, current %s",
+            lastFired,
+            timer.getTimestamp());
+        if (lastFired != null && lastFired.isBefore(timer.getTimestamp())) {
+          pushedBackTimers.add(timer);
```

Comment:
Now that I look more closely, doesn't this cause the timers to fire one at a time if they are in order, with all the rest pushed back into the commit result? And only when a timer is moved earlier you go to the other branch and fire multiple timers in the same bundle.

Reply:
Generally yes. The problem is that you cannot fire multiple timers at once, because the outcome of firing a single timer is unknown. It might or might not setup some other timer. If it was to be solved, it seems to me it would require `PushbackSideInputDoFnRunner` (and most likely `DoFnRunner` to return set timers, instead of `void`. The other option would be somehow hack into `WatermarkManager` and be able to retrieve "most recent timers set", os something like that (which seems to be fragile and hackish). The correct solution seems to be a too big refactor, given that the performance penalty for the solution proposed in this PR seems to be low.